### ISSUE TITLE
Make DropoutLoss honor `reduction` so it composes with MultiLoss/MultitaskLoss

### ISSUE THREAD

**Issue by dahvida** [emdgroup/matcha#59]
## Summary

`DropoutLoss` (and its `dropout-*` aliases) constructs successfully inside `MultiLoss` / `MultitaskLoss` after the fix in `028ecf4` (#55), but training still crashes on the first batch with `too many indices for tensor of dimension 0`. The wrapper losses hand each inner loss a per-element tensor and mask/reduce themselves, but `DropoutLoss.forward` unconditionally returns a scalar. This is the deferred shape-contract half of #55 called out in the regression test docstring.

## Current Behavior

`pretrain_multitask` runs with a `MultiLoss` config mixing `dropout-mse` (descriptors branch) and `dropout-focal-bce` (fingerprints branch) die at `Epoch 0: 0/22572` with:

```
Error running command 'pretrain_multitask': too many indices for tensor of dimension 0
```

Root cause (`src/matcha/nn/losses.py`):

1. `MultiLoss.__init__` (losses.py:167) instantiates each inner loss as `LossRegistry[loss_fn](reduction="none", **loss_args)`, expecting a per-element tensor from the inner forward. `MultitaskLoss.__init__` (losses.py:111) does the same.
2. `DropoutLoss.__init__` (losses.py:626, added in `028ecf4`) silently pops the caller-supplied `reduction` and always passes `reduction="none"` only to the *inner* per-element loss.
3. `DropoutLoss.forward` (losses.py:659) always reduces to a scalar: `losses.sum() / (keep_mask.sum() + 1e-8)`.
4. `MultiLoss.forward` (losses.py:228) then does `task_loss[valid_mask] = 0.0` on the returned 0-d tensor and raises.

The failure is encoder-agnostic — GT and GatedGCN die identically with the same config family. Before `028ecf4` the same runs died at construction with the duplicated-`reduction` TypeError; the fix moved the failure from init to first forward.

`028ecf4`'s regression test explicitly documents this deferral:

> "Only construction is verified: MultitaskLoss's forward expects per-element inner output, but DropoutLoss reduces internally — a separate shape-contract issue (see issue 55 plan)."

## Proposed Behavior

`DropoutLoss` should honor a `reduction` parameter like every other loss in `LossRegistry`. When `reduction="none"`, `forward` should return the masked per-element loss tensor (still zeroing dropped and NaN entries, no scalar reduction). When `reduction="mean"` (default) or `"sum"`, it should reduce as today. This restores the wrapper composition contract: `MultiLoss` and `MultitaskLoss` pass `reduction="none"` at construction and expect per-element output back, and the eight `dropout-*` aliases inherit the fix through `**kwargs` forwarding.

Standalone use (`DropoutMSELoss(dropout=0.5, seed=0)(...)`) must keep returning a scalar unchanged, since the default remains `reduction="mean"`.

## Acceptance Criteria

- Running `pretrain_multitask` with a `MultiLoss` config mixing `dropout-mse` and `dropout-focal-bce` trains past the first batch without a `too many indices for tensor of dimension 0` error.
- `DropoutLoss` accepts a `reduction` parameter (`"none"`, `"mean"`, `"sum"`) with `"mean"` as default.
- `DropoutLoss(reduction="none")(...)` returns a tensor with the same shape as its inputs, with dropped and NaN entries zeroed. `DropoutLoss(reduction="mean")(...)` returns a scalar equal to today's output. `DropoutLoss(reduction="sum")(...)` returns the sum of the masked per-element loss.
- All eight `dropout-*` aliases inherit the new `reduction` behavior via `**kwargs` forwarding — construction with `reduction="none"` and per-element output verified for at least `dropout-mse`, `dropout-mae`, `dropout-focal-bce`, `dropout-weighted-bce`.
- A regression test in `tests/nn/test_losses.py` covers `MultiLoss(loss_configs=[{"loss_fn": "dropout-mse", ...}, {"loss_fn": "dropout-focal-bce", ...}])` **end-to-end through forward** on a small batch with NaN targets, asserting no exception and a finite scalar total loss. Analogous test for `MultitaskLoss(loss_fn="dropout-mse", ...)`.
- Existing regression tests from `028ecf4` (construction-only assertions) continue to pass.

## Context

- Follows #55 / `028ecf4` (constructor collision fix). The plan for that issue explicitly deferred the forward-composition problem to a follow-up.
- Reproducer: `pretrain_multitask` with a `MultiLoss` config that mixes `dropout-mse` (descriptors branch) and `dropout-focal-bce` (fingerprints branch) — the same shape as the config in issue #55. Two independent SLURM jobs on our cluster (`pt-gt-feat`, `pt-gatedgcn-feat`) both died at Epoch 0 batch 0 on 2026-08-14. The failure is encoder-agnostic: GT and GatedGCN produce the same traceback with the same config family.

## Technical Notes

- Fix candidate: add `reduction: str = "mean"` to `DropoutLoss.__init__`, store on `self`, and branch in `forward`:
  - Currently `forward` computes `losses = self.loss(outputs, targets)`, `keep_mask = ~nan_mask & keep_from_dropout`, then `return (losses * keep_mask).sum() / (keep_mask.sum() + 1e-8)`.
  - For `reduction="none"`: return `losses * keep_mask` (dropped/NaN entries zeroed, shape preserved) — this is what `MultiLoss.forward` then masks and averages.
  - For `reduction="sum"`: return `(losses * keep_mask).sum()`.
  - For `reduction="mean"`: keep today's behavior.
- The `kwargs.pop("reduction", None)` added in `028ecf4` should be replaced by an explicit `reduction` parameter on `DropoutLoss.__init__`; the inner loss keeps `reduction="none"` hard-coded (that invariant is still owned by `DropoutLoss`).
- Consider whether `MultiLoss.forward`'s post-inner masking (losses.py:228 `task_loss[valid_mask] = 0.0`) is still meaningful when the inner loss is `DropoutLoss` — the NaN mask has already been applied inside `DropoutLoss`, so the double-mask is harmless but redundant. Not a blocker; note only.
- Relevant files: `src/matcha/nn/losses.py`, `tests/nn/test_losses.py`.

**Comment by dahvida** [comment]
## Implementation Plan — Make `DropoutLoss` honor `reduction`

This comment is the plan-of-record for issue #59 and will guide staged implementation. The scope is a focused bug fix, so the plan is a single stage.

### Overview

`DropoutLoss` currently discards any caller-supplied `reduction` kwarg and always returns a scalar, breaking composition with `MultiLoss`/`MultitaskLoss` (both of which pass `reduction="none"` and expect a per-element tensor). We add an explicit `reduction` parameter to `DropoutLoss.__init__`, branch in `forward`, and cover the composition path end-to-end with tests. All 8 `dropout-*` aliases inherit the fix automatically via `**kwargs` forwarding. Default `reduction="mean"` preserves current standalone behavior.

The approach mirrors the pattern already used in the same file by `BCEFocalLoss`, `Poly1BCELoss`, and `WeightedBCELoss`: take `reduction` in `__init__`, store on `self`, branch in `forward` between `.mean()`, `.sum()`, and returning the per-element tensor. Unknown `reduction` values silently fall through to per-element output (matches file convention). The inner loss stays hardcoded to `reduction="none"` — that invariant is still owned by `DropoutLoss`.

### Stage 1 — Honor `reduction` in `DropoutLoss` and add end-to-end wrapper tests

**Goal**: Make `DropoutLoss` return per-element losses when constructed with `reduction="none"`, so `MultiLoss` and `MultitaskLoss` composition works end-to-end. Preserve default scalar behavior for standalone use.

**Files modified**:

- `src/matcha/nn/losses.py`
  - `DropoutLoss.__init__` (currently losses.py:603–627):
    - Add `reduction: str = "mean"` parameter to the signature.
    - Store `self.reduction = reduction`.
    - Delete the `kwargs.pop("reduction", None)` line — superseded by the explicit parameter.
    - Inner loss still instantiated with `reduction="none"` hardcoded (invariant retained).
  - `DropoutLoss.forward` (currently losses.py:629–659):
    - Compute `losses = losses * keep_mask` as today.
    - Branch on `self.reduction`:
      - `"mean"` → return `losses.sum() / (keep_mask.sum() + 1e-8)` (today's behavior).
      - `"sum"` → return `losses.sum()`.
      - else (including `"none"`) → return the per-element `losses` tensor (dropped and NaN entries already zeroed by the multiplication).
  - Docstring updates:
    - Class docstring: document the new `reduction` parameter and its interaction with dropout masking.
    - `forward` docstring: update the returns section to describe shape/scalar depending on reduction mode.

- `tests/nn/test_losses.py` (additions to `TestDropoutLoss` and `TestDropoutAliases`)
  - **Unit tests on `reduction` branches** (added to `TestDropoutLoss`):
    - `test_reduction_mean_matches_current_behavior`: `DropoutLoss(loss_fn="mse", dropout=0.0)` (default `reduction="mean"`) returns a scalar equal to `F.mse_loss(preds, targets, reduction="mean")`. Guards against regression in the standalone default.
    - `test_reduction_none_returns_per_element`: `DropoutLoss(loss_fn="mse", reduction="none", dropout=0.0)` in eval mode returns a tensor of shape `(batch, num_tasks)` matching `F.mse_loss(preds, targets, reduction="none")`.
    - `test_reduction_sum_matches_masked_sum`: `DropoutLoss(loss_fn="mse", reduction="sum", dropout=0.0)` returns a scalar equal to `F.mse_loss(preds, targets, reduction="sum")` (no dropout, no NaN).
    - `test_reduction_none_zeros_nan_and_dropout`: with dropout > 0 in train mode + NaN targets, `reduction="none"` returns per-element tensor whose NaN positions and dropped positions are exactly 0.
    - `test_reduction_none_preserves_shape`: output shape matches input `(batch, num_tasks)` for `reduction="none"`.
  - **End-to-end wrapper regression tests** (added to `TestDropoutLoss`):
    - `test_wrapped_by_multi_loss_forward`: build a `MultiLoss` with a realistic mixed config — one entry using `dropout-mse` on columns `[0,1,2]` and one using `dropout-focal-bce` on columns `[3,4]`. Run forward on a `(batch, 5)` batch with sprinkled NaN targets. Assert no exception, `total_loss` is a finite scalar, and `loss_log` contains both task entries.
    - `test_wrapped_by_multitask_loss_forward`: `MultitaskLoss(loss_fn="dropout-mse", loss_args={"dropout": 0.5, "seed": 0})` forward pass on the `multitask_targets` fixture. Assert no exception and finite scalar output. Seed is pinned for reproducibility.
    - Keep existing `test_wrapped_by_multitask_loss_constructs` / `test_wrapped_by_multi_loss_constructs` (construction-only) unchanged so they continue to pass. Update their docstrings to drop the deferral note (issue now resolved).
  - **Alias-level per-element output** (added to `TestDropoutAliases`):
    - `test_aliases_reduction_none_returns_per_element` parametrized over `["dropout-mse", "dropout-mae", "dropout-focal-bce", "dropout-weighted-bce"]`: instantiate with `reduction="none"`, dropout=0.0, forward on a small `(batch, 1)` batch, assert output is a tensor with the same shape as inputs (2-D, not 0-D).

**Testing approach**:
- Framework: pytest (matches project convention).
- Test types: unit tests for reduction branches + integration/regression tests for wrapper composition (`MultiLoss`, `MultitaskLoss` forward end-to-end).
- Coverage target: patch coverage ≥ 80% per CONTRIBUTING.md; adding ~8 focused tests covers every new branch in `DropoutLoss.forward` plus the wrapper composition paths.
- Run locally with `uv run python -m pytest -k 'not gpu' tests/nn/test_losses.py`.

**Verification**:
- All existing `TestDropoutLoss`, `TestDropoutAliases`, `TestMultiLoss`, `TestMultitaskLoss` tests continue to pass.
- New tests assert every acceptance criterion from the issue body.
- Pre-commit hooks (Ruff, Pyright) pass on the change.

**Commit type**: `fix` (PATCH bump).

**Dependencies**: none — self-contained; builds on the constructor fix from #56 / `028ecf4` that is already merged to `main`.

**Scope check**: Change touches one class (`DropoutLoss`) in one source file plus additions to two test classes. Roughly 15 lines of source changes plus ~80 lines of tests. Well within a single Claude Code CLI session.

## Decision Log

### Clarifying Questions (Step 3)

- **Q: How should `DropoutLoss` handle unknown `reduction` values?**
  A: Silent fallback to per-element output — matches the convention already used by `BCEFocalLoss`, `Poly1BCELoss`, and `WeightedBCELoss` in the same file. Consistency over strictness.

- **Q: For the end-to-end regression tests, which reproducer should the test mirror?**
  A: Realistic reproducer — `MultiLoss` with a mixed `dropout-mse` + `dropout-focal-bce` config on multi-column targets with NaN. Mirrors the exact `pretrain_multitask` configuration that failed on 2026-08-14 in the two SLURM jobs called out in the issue body.

- **Q: Should `MultiLoss.forward`'s post-inner masking (losses.py:228 `task_loss[valid_mask] = 0.0`) be refactored to skip the double-mask when the inner is `DropoutLoss`?**
  A: Leave as-is. The technical notes on the issue confirm the double-mask is "harmless but redundant" — refactoring it would expand scope beyond the shape-contract fix and risk unrelated loss compositions. Filed only as a note.

### Architecture Choice (Step 4)

**Selected: Approach A — explicit `reduction` parameter with branch in `forward`.**

Rationale:
- Matches the existing convention in `losses.py` (`BCEFocalLoss`, `Poly1BCELoss`, `WeightedBCELoss`).
- Change surface is strictly bounded to `DropoutLoss` — no cross-cutting refactor.
- Default `reduction="mean"` keeps every one of the 14 existing `TestDropoutLoss` tests passing unchanged, so we get a clean before/after regression contract.
- Mirrors the fix candidate in the issue's Technical Notes verbatim.

Alternative considered and rejected:
- **Approach B — extract a shared `_reduce()` helper and refactor all four losses in the file that hand-roll the branch.** DRY improvement, but expands scope well beyond the bug this issue targets and would touch three unrelated losses. Filed as a plausible follow-up but not appropriate to ride on a fix.

---

This comment will guide the staged implementation. Use `/mach10:issue-implement 59 1` in a fresh session to begin Stage 1.

**Comment by dahvida** [comment]
### Stage 1 progress — `DropoutLoss.reduction` honored

Commit: `67b129e`

**Source (`src/matcha/nn/losses.py`)**
- `DropoutLoss.__init__` now takes an explicit `reduction: str = "mean"` and stores it on `self.reduction`. Removed the `kwargs.pop("reduction", None)` line — superseded by the explicit parameter.
- `DropoutLoss.forward` branches after masking: `"mean"` returns `losses.sum() / (keep_mask.sum() + 1e-8)` (today's default, preserved for standalone use), `"sum"` returns `losses.sum()`, else (including `"none"`) returns the per-element `losses` tensor with dropped/NaN entries already zeroed by the multiplication.
- Inner loss stays hardcoded to `reduction="none"` — invariant retained.
- Class + `forward` docstrings updated to describe the new parameter and return shape.

**Tests (`tests/nn/test_losses.py`)**
- Added 5 unit tests to `TestDropoutLoss` covering the three reduction branches, NaN/dropout zeroing under `reduction="none"`, and shape preservation.
- Added `test_wrapped_by_multi_loss_forward` — end-to-end regression mirroring the `pretrain_multitask` config family from the issue body (`dropout-mse` on descriptor columns + `dropout-focal-bce` on fingerprint columns, sprinkled NaNs). Asserts finite scalar total loss and both named task entries in the log.
- Added `test_wrapped_by_multitask_loss_forward` — analogous forward regression for `MultitaskLoss(loss_fn="dropout-mse", ...)`.
- Added `test_aliases_reduction_none_returns_per_element` to `TestDropoutAliases`, parametrized over `dropout-mse`, `dropout-mae`, `dropout-focal-bce`, `dropout-weighted-bce`.
- Dropped the deferral note from the existing construction-only `test_wrapped_by_multitask_loss_constructs` / `test_wrapped_by_multi_loss_constructs` docstrings (they still assert construction; forward is now covered by the two new tests).

**Verification**
- `uv run python -m pytest -k 'not gpu' tests/nn/test_losses.py` — **134 passed**.
- `pre-commit` (ruff, ruff-format, pyright) — clean on both modified files.

No deviations from the plan.
